# Defaults
molgenisStandInScript='/groups/umcg-fg/tmp01/projects/pgx-passport/tools/PGx-pipeline/molgenis_stand_in_script.sh'
pythonVersion='Python/3.9.1-GCCcore-7.3.0-bare'
pythonEnvironment='/groups/umcg-fg/tmp01/projects/pgx-passport/tools/virtualenv/3.9.1-GCCcore-7.3.0-bare'
optiCallExecutable='/groups/umcg-fg/tmp01/projects/pgx-passport/tools/tss101-opticall-090d8ed4136a/opticall/opticall'
RPlusVersion='RPlus/4.0.3-foss-2018b-v21.01.1'
plinkVersion='PLINK/1.9-beta6-20190617'
plink2Version='PLINK/2.0-alpha4.5-20230813'
htslibVersion='HTSlib/1.11-GCCcore-7.3.0'
boostVersion='Boost/1.67.0-foss-2018b'
shapeitExecutable='/groups/umcg-fg/tmp01/projects/pgx-passport/tools/shapeit4-4.2.1/bin/shapeit4.2'
variantsPassedQualityControl='/groups/umcg-fg/tmp01/projects/pgx-passport/data/GSAMD-24v3/passed_variants_qc.txt'
javaVersion='Java/11-LTS'
nextflowPath='/groups/umcg-fg/tmp01/projects/pgx-passport/tools/nextflow'
imputationPipelineReferencePath='/groups/umcg-fg/tmp01/projects/pgx-passport/data/processed/imputationreference'
outputName='genimpute_phased_imputed'
imputationFlankSize='1000000'
pipelineRoot='/groups/umcg-fg/tmp01/projects/pgx-passport/tools/PGx-pipeline'
asterixRoot='/groups/umcg-fg/tmp01/projects/pgx-passport/tools/asterix'
rLibsPath='/groups/umcg-fg/tmp01/projects/pgx-passport/tools/Rlibs'
bpmFile='/apps/data/GSAarray/GSAMD-24v3-0-EA_20034606_A1.bpm'
cnvConfig='/groups/umcg-fg/tmp01/projects/pgx-passport/tools/PGx-pipeline/data/cyp2d6/config.yml'
cnvBedFile='/groups/umcg-fg/tmp01/projects/pgx-passport/tools/PGx-pipeline/data/cyp2d6/cyp2d6.bed'
pgxGenesBed37='/groups/umcg-fg/tmp01/projects/pgx-passport/tools/PGx-pipeline/data/pgx-genes_GRCh37.bed'
pgxGenesBed38='/groups/umcg-fg/tmp01/projects/pgx-passport/tools/PGx-pipeline/data/pgx-genes_GRCh38.bed'
pgxGenesBed38Flanked='/groups/umcg-fg/tmp01/projects/pgx-passport/tools/PGx-pipeline/data/pgx-genes_GRCh38_1Mb-flanks.bed'
gnomadAnnotationFile='/groups/umcg-fg/tmp01/projects/pgx-passport/results/gnomad/analyses/concatenated_pgx_genes/out/pgx_genes_gnomad_chromosomes_renamed.vcf.gz'
correctiveVariantsOutputDir='/groups/umcg-gdio/tmp01/projects/2021001/pgx-pipeline/analyses/select_corrective_variants/out'

# Parameters specific for each run
workdir=
outdir=
samplesheet=
cnvOutDir="${outdir}/cyp2d6_cnv_calling/out"
gtcDataDir=
finalReportsDir="${outdir}/final_reports/out"
stagedIntensities="${outdir}/staged_intensities/out"
sampleListPrefix="${outdir}/quality_controlled_samples/out/sample_list"
_genotypesOxfordPrefix="${outdir}/oxford_gen_sample/chr_${chromosomeNumber}"
_genotypesPlinkPrefix='${outdir}/qced_plink_files/chr_${chromosomeNumber}'
concatenatedGenotypesOutputDir="${outdir}/concatenated_genotypes/out"
imputationOutputDir="${outdir}/impute_pgx_genes/out/"
